docs/topics/concurrency.rst

```diff
@@ -70,40 +75,37 @@ Here is the program in examples/thread_pool_executor.py.
         reversed.
         """
 
-        with rasterio.Env():
-
-            with rasterio.open(infile) as src:
-
-                # Create a destination dataset based on source params. The
-                # destination will be tiled, and we'll process the tiles
-                # concurrently.
-                profile = src.profile
-                profile.update(blockxsize=128, blockysize=128, tiled=True)
-
-                with rasterio.open(outfile, "w", **profile) as dst:
-
-                    # Materialize a list of destination block windows
-                    # that we will use in several statements below.
-                    windows = [window for ij, window in dst.block_windows()]
-
-                    # This generator comprehension gives us raster data
-                    # arrays for each window. Later we will zip a mapping
-                    # of it with the windows list to get (window, result)
-                    # pairs.
-                    data_gen = (src.read(window=window) for window in windows)
-
-                    with concurrent.futures.ThreadPoolExecutor(
-                        max_workers=num_workers
-                    ) as executor:
-
-                        # We map the compute() function over the raster
-                        # data generator, zip the resulting iterator with
-                        # the windows list, and as pairs come back we
-                        # write data to the destination dataset.
-                        for window, result in zip(
-                            windows, executor.map(compute, data_gen)
-                        ):
-                            dst.write(result, window=window)
+        with rasterio.open(infile) as src:
+
+            # Create a destination dataset based on source params. The
+            # destination will be tiled, and we'll process the tiles
+            # concurrently.
+            profile = src.profile
+            profile.update(blockxsize=128, blockysize=128, tiled=True)
+
+            with rasterio.open(outfile, "w", **src.profile) as dst:
+                windows = [window for ij, window in dst.block_windows()]
+
+                # We cannot write to the same file from multiple threads
+                # without causing race conditions. To safely read/write
+                # from multiple threads, we use a lock to protect the
+                # DatasetReader/Writer
+                read_lock = multiprocessing.Lock()
+                write_lock = multiprocessing.Lock()
```

Comment:
Should these be `threading.Lock`?

Reply:
You are totally right, I was testing around with the `ProcessPoolExecutor`, after that I forgot to change that back.